Using Convolutional Neural Networks to predict if an image is a Cat or a Dog

Importing the libraries and packages from keras to implement Convolutional Neural Networks

In [1]:
from keras.models import Sequential
from keras.layers import Conv2D
from keras.layers import MaxPooling2D
from keras.layers import Flatten
from keras.layers import Dense

Using TensorFlow backend.


Initializing the CNN

In [2]:
classifier = Sequential()

Step 1 - Convolution

In [3]:
classifier.add(Conv2D(32, (3, 3), input_shape = (64, 64, 3), activation = 'relu'))

Step 2 - Pooling

In [4]:
classifier.add(MaxPooling2D(pool_size = (2, 2)))

Adding a second Convolution Layer

In [5]:
classifier.add(Conv2D(32, (3, 3), activation = 'relu'))
classifier.add(MaxPooling2D(pool_size = (2, 2)))

Step 3 - Flattening

In [7]:
classifier.add(Flatten())

Step 4 - Full Connection

In [8]:
classifier.add(Dense(units = 128, activation = 'relu'))
classifier.add(Dense(units = 1, activation = 'sigmoid'))

Compiling the CNN

In [9]:
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

Fitting the CNN to our images

In [10]:
from keras.preprocessing.image import ImageDataGenerator

Image Augmentation as part of image preprocessing

In [11]:
train_datagen = ImageDataGenerator(rescale = 1./255,
                                   shear_range = 0.2,
                                   zoom_range = 0.2,
                                   horizontal_flip = True)

test_datagen = ImageDataGenerator(rescale = 1./255)

Initializing the Training Set

In [13]:
training_set = train_datagen.flow_from_directory('dataset/training_set',
                                                 target_size = (64, 64),
                                                 batch_size = 32,
                                                 class_mode = 'binary')

Found 8000 images belonging to 2 classes.


Initializing the Test Set

In [14]:
test_set = test_datagen.flow_from_directory('dataset/test_set',
                                            target_size = (64, 64),
                                            batch_size = 32,
                                            class_mode = 'binary')

Found 2000 images belonging to 2 classes.


Fitting the classifier to the training set

In [16]:
classifier.fit_generator(training_set,
                         steps_per_epoch = 8000,
                         epochs = 1,
                         validation_data = test_set,
                         validation_steps = 2000)

Epoch 1/1
8000/8000 [==============================] - 9972s 1s/step - loss: 0.3714 - acc: 0.8267 - val_loss: 0.4517 - val_acc: 0.8182


An accuracy of 82.67 % on Training Data and 81.82 % on test data.

Predicting if a given image is a Dog or a Cat; cat_or_dog_1 is a Dog and cat_or_dog_2 is a Cat

In [19]:
import numpy as np
from keras.preprocessing import image
test_image = image.load_img('dataset/single_prediction/cat_or_dog_2.jpg', target_size = (64, 64))
test_image = image.img_to_array(test_image)
test_image = np.expand_dims(test_image, axis = 0)
result = classifier.predict(test_image)

In [21]:
result

array([[ 0.]], dtype=float32)

In [22]:
training_set.class_indices

{'cats': 0, 'dogs': 1}

The value of result variable is 0 which implies the above image is classified as a CAT

In [23]:
if result[0][0] == 1:
    prediction = 'dog'
else:
    prediction = 'cat'

In [20]:
prediction

'cat'